kubeadm: document alpha addon-installer init phase/feature-gate (#18023)
kubernetes/website: content/en/docs/reference/setup-tools/kubeadm/kubeadm-init.md

Please note that the configuration field `kubernetesVersion` or the command line flag
`--kubernetes-version` affect the version of the images.

### Installing custom addons (Alpha) {#addon-installer}

> 

Reviewer: Add a [feature-state shortcode](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#feature-state) here.

> Kubeadm will normally install CoreDNS/kube-dns and kube-proxy using manifests built into the kubeadm binary.

Reviewer: ```suggestion
Normally, kubeadm deploys two [add-ons](/docs/concepts/cluster-administration/addons/), CoreDNS/kube-dns and kube-proxy, using manifests built into kubeadm itself.
```

> Using the `AddonInstaller` feature-gate (alpha), you can instruct kubeadm to install addons from other sources.

Reviewer: ```suggestion
You can instruct kubeadm to install addons from other sources if you enable the `AddonInstaller` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/).
```

> Kubeadm will no longer use the in-binary manifests when using the `AddonInstaller` feature-gate.

Reviewer: ```suggestion
If you enable the the `AddonInstaller` feature gate, kubeadm no longer uses its built-in add-on manifests.
```
How does this fit in with the explanation on line 164?

> This feature is driven by a Component Config called the `addons.config.x-k8s.io/AddonInstallerConfiguration`.

Reviewer: ```suggestion
This feature is driven by a component configuration: `addons.config.x-k8s.io/AddonInstallerConfiguration`.
```

The `AddonInstallerConfiguration` can be specified in the kubeadm config-file.

> If the `AddonInstaller` feature-gate is specified, but a configuration is not supplied, kubeadm will default the API. It is planned to default to installing version appropriate CoreDNS+kube-proxy, but these addons are not yet implemented.

Reviewer: I recommend:
```suggestion
If you enable the `AddonInstaller` feature gate, but you don't specify a configuration, kubeadm uses its own default configuration.
```
([avoid making statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future))

Minimal usage in a kubeadm config could look like:
```yaml
apiVersion: kubeadm.k8s.io/v1beta2
kind: InitConfiguration
---
apiVersion: kubeadm.k8s.io/v1beta2
kind: ClusterConfiguration
featureGates:
  AddonInstaller: true
```
```shell
kubeadm init --config cluster1-config.yaml
```

Supplying your own configuration, you can override the feature's default and install from a manifest-list or kustomize directory:
```yaml
apiVersion: kubeadm.k8s.io/v1beta2
kind: InitConfiguration
---
apiVersion: kubeadm.k8s.io/v1beta2
kind: ClusterConfiguration
featureGates:
  AddonInstaller: true
---
apiVersion: addons.config.x-k8s.io/v1alpha1
kind: AddonInstallerConfiguration
addons:
- name: multibase-pod
  kustomizeRef: github.com/kubernetes-sigs/kustomize//examples/multibases/dev/?ref=v1.0.6
- name: weavenet
  manifestRef: https://cloud.weave.works/k8s/v1.10/net.yaml
```
```shell
kubeadm init --config cluster1-config.yaml
```

You may also invoke the addon-installer phase independently to apply the configuration without upgrading the rest of the cluster:
```shell
kubeadm init addons installer --config cluster1-config.yaml
```

> The initial KEP for this feature can be found here: [kubeadm-addon-installer-KEP](TODO: link KEP merge location)

Reviewer: This isn't needed; I would omit it.
